### ISSUE TITLE
anchor-is-valid should not throw an error when using `javascript:void(0)` as a value

### ISSUE THREAD

**Issue by nicklemmon** [evcohen/eslint-plugin-jsx-a11y#524]
```
<a href="javascript:void(0);" role="button">Click Me</a>
```

Is valid when an element styled as a link is needed to behave as a button. Hashes, on the other hand, are not valid. I believe an exception should be carved out for this particular value.

Thanks!

**Comment by nicklemmon** [comment]
Oddly enough in that same thread, a user has a clever solution by allowing:
`href="void"` that is handled globally. Either way, using an href is a possibility!

**Comment by ljharb** [comment]
That you *can* do it doesn't mean you *should*.

I agree that `javascript:void(0);` is better than `#`, but neither are good. The proper thing if you want to "not navigate somewhere" is to use "not a link", and I'm not really sure what arguments there are to contradict that.

**Comment by nicklemmon** [comment]
Well, I would say that it's more complex than that, and often has to do with the interplay of styling and semantics. Technically adding `role="button"` makes it just as 'not a link' (to use your words) as `<button>` though then one is left with how to handle presentation.

Say, for example, there is a `<p>` with an element inside that opens a modal. That, semantically, must be a button. A `<button>` doesn't wrap too nicely inside a `<p>` so then one would typically use an `<a>` particularly if said element should look like the other things that are actual links within the paragraph.

Now the solution could be to have a class that makes any type of element, regardless of semantics, look like a link. Alternatively, you just have styles applied to the `<a>` that are global and then one adds `role="button"` and some valid `href` to boot and then there isn't an additional class to manage.

...and all of that is informed by how one is styling their components, which in 2019, can mean a whole lot of different things.

All I'm saying is, it isn't a cut and dry thing and most things in the a11y world aren't, which is why linting for problems is so difficult. Feel free to close this as it seems like I am not making much progress persuading you, though I hope the discussion is valuable.

I appreciate your work on this project. 👍

**Comment by ljharb** [comment]
I’d say that if it looks like a link, it should navigate somewhere :-) also consider the scenario where JavaScript breaks or fails to load - you can disable JS to simulate this common failure scenario.

**Comment by ljharb** [comment]
Indeed :-) and I’d strongly encourage server rendering regardless.

I’m going to close this; thanks for the discussion. Definitely feel free to disable the rule (inline or overall) if you don’t like how it works.
